crypto/cmp/cmp_client.c

```diff
@@ -0,0 +1,878 @@
+ * is before the "checkAfter" sent by the server. If ctx->totaltimeout is 0, the
+ * timeout is disabled.
+ *
+ * returns 1 on success, returns received PKIMESSAGE in *msg argument
+ * returns 0 on error or when timeout is reached without a received message
+ *
+ * TODO: handle multiple poll requests for multiple certificates
+ *       --> GitHub issue#67
+ */
+static int pollForResponse(OSSL_CMP_CTX *ctx, int rid, OSSL_CMP_MSG **out)
+{
+    OSSL_CMP_MSG *preq = NULL;
+    OSSL_CMP_MSG *prep = NULL;
+    OSSL_CMP_POLLREP *pollRep = NULL;
+
+    ossl_cmp_info(ctx,
+                  "received 'waiting' PKIStatus, starting to poll for response");
+    for (;;) {
+        if ((preq = ossl_cmp_pollReq_new(ctx, rid)) == NULL)
+            goto err;
+
+        if (!send_receive_check(ctx, preq, &prep, OSSL_CMP_PKIBODY_POLLREP,
+                                CMP_R_POLLREP_NOT_RECEIVED))
+            goto err;
+
+        /* handle potential pollRep */
+        if (ossl_cmp_msg_get_bodytype(prep) == OSSL_CMP_PKIBODY_POLLREP) {
+            int64_t check_after;
+            OSSL_CMP_POLLREPCONTENT *prc = prep->body->value.pollRep;
+
+            /*
+             * TODO: handle multiple elements, in case multiple requests have
+             * been sent - see https://github.com/mpeylo/cmpossl/issues/67
+             */
+            if (sk_OSSL_CMP_POLLREP_num(prc) > 1) {
+                CMPerr(0, CMP_R_MULTIPLE_RESPONSES_NOT_SUPPORTED);
+                goto err;
+            }
+            if ((pollRep = ossl_cmp_pollrepcontent_get0_pollrep(prc, rid))
+                == NULL)
+                goto err;
+            if (!ASN1_INTEGER_get_int64(&check_after, pollRep->checkAfter)) {
+                CMPerr(0, CMP_R_BAD_CHECKAFTER_IN_POLLREP);
+                goto err;
+            }
+            if (check_after < 0) {
+                CMPerr(0, CMP_R_RECEIVED_NEGATIVE_CHECKAFTER_IN_POLLREP);
+                goto err;
+            }
+            /* TODO: print OPTIONAL reason (PKIFreeText) from message */
+            ossl_cmp_log1(INFO, ctx,
+                          "received polling response, waiting for check_after = %ld seconds",
+                          check_after);
+
+            if (ctx->totaltimeout > 0) { /* total timeout is not infinite */
+                const int exp = 5; /* expected max time per msg round trip */
+                int64_t time_left = (int64_t)(ctx->end_time - exp - time(NULL));
+
+                if (time_left <= 0) {
+                    CMPerr(0, CMP_R_TOTAL_TIMEOUT);
+                    goto err;
+                }
+                if (time_left < check_after)
+                    check_after = time_left;
+                /* poll one last time just when timeout was reached */
+            }
+
+            OSSL_CMP_MSG_free(preq);
+            preq = NULL;
+            OSSL_CMP_MSG_free(prep);
+            prep = NULL;
+            sleep((unsigned int)check_after);
```

Comment:
Not sure I like a sleep here. It provides no opportunity for the application to do async processing and go off and do something else while it is waiting.

Reply:
Unfortunately the way the API is designed so far there is not much choice.
I'll try coming up with an alternative that does not change too much...

Reply:
I plan to add a more flexible CMP cert enrollment function:
```
int OSSL_CMP_try_certreq(OSSL_CMP_CTX *ctx, int req_type, int wait);
```
Its input is the CMP context as usual, the type of cert request (ir/cr/p10cr/kur), 
and the remaining time to wait after a previous call, or `-1` for the initial call.
It returns `0` as soon as the requested cert is available (then the application can retrieve it using `OSSL_CMP_CTX_get0_newCert()`)
or else the amount of time to wait before trying again.
In the latter case the application can decide to sleep for some time and/or do whatever before retrying. Then it calls the same function again, giving any remaining time to wait before trying again or `0` for immediate retry.

@mattcaswell, how does this sound?

Reply:
I've meanwhile conceived the following enhanced API design with improved usability and flexibility, which I'm going to implement today:
```
int OSSL_CMP_try_certreq(OSSL_CMP_CTX *ctx, int req_type);
```
Its input is the CMP context as usual and the type of cert request (ir/cr/p10cr/kur).
It returns 0 on error or -1 if the requested cert is available. Thenthe application can use `OSSL_CMP_CTX_get0_newCert()` to retrieve the new certificate (which will yield `NULL` in case an error occurred).
If no error occurred but no certificate is available yet then `OSSL_CMP_try_certreq()` remembers in the CMP context the point in time whan polling should be retried and returns a positive value indicating the number of seconds the application should let pass before trying again. In this case the application can decide to sleep for some time and/or do anything else before retrying. Then it calls the same function again with the same parameter values as before. In this case `OSSL_CMP_try_certreq()` sleeps until the point in time remembered has passed and then polls to see if meanwhile the requested certificate is available.
If the application decides to abort its pending certificate request it can use a different request type (or any other value, such as -1). In this case `OSSL_CMP_try_certreq()` aborts the CMP transaction by sending an error message to the server.

@mpeylo, I hope you also like this solution?

Reply:
This sounds like a cleaner solution.

Reply:
Pleased to hear. After discussion with a colleague I've slightly revised the design further according to my edit/update of the above description (because the application should be able to overrule/ignore the check-after time provided by the server).

Reply:
@mattcaswell and @Akretsch, finally the new function `OSSL_CMP_try_certreq()`, including documentation and tests, is ready for review.